### PR TITLE
Add key size validation for AddEncryptionKey

### PR DESCRIPTION
**Pull request by NoahStahl**
To address #1151 

I believe for the symmetric key case we just need to check the KeySize property which is expected to be 256 based on that as the supported algorithm. I added an exception message and associated tests following the existing pattern.

I'm not sure what should be expected with the asymmetric case as to the key size, so no changes made there.

### CODE REVIEW

File: src/OpenIddict.Abstractions/Resources/OpenIddictResources.resx [hunk 1]
```diff
@@ -1385,6 +1385,9 @@ To register the OpenIddict core services, reference the 'OpenIddict.Core' packag
     <value>The implicit flow must be enabled when adding a response type containing '{0}'.</value>
     <comment>{Locked}</comment>
   </data>
+  <data name="ID0283" xml:space="preserve">
+    <value>Provided symmetric key was incorrect size. Must be 256 bits derived from exactly 32 bytes or string of length 32.</value>
```
Comment (kevinchalet) [inline src/OpenIddict.Abstractions/Resources/OpenIddictResources.resx]: Can you please make it parametrized? Also, I'd remove the `derived from exactly 32 bytes or string of length 32` part, which is a bit confusing.

File: src/OpenIddict.Abstractions/Resources/OpenIddictResources.resx [hunk 2]
```diff
@@ -1385,6 +1385,9 @@ To register the OpenIddict core services, reference the 'OpenIddict.Core' packag
...
+  </data>
```
Comment (kevinchalet) [inline src/OpenIddict.Abstractions/Resources/OpenIddictResources.resx:1391]: Please add `<comment>{Locked}</comment>` to ensure it's not localized.

File: src/OpenIddict.Server/OpenIddictServerBuilder.cs [hunk 3]
```diff
@@ -175,6 +177,11 @@ public OpenIddictServerBuilder AddEncryptionKey(SecurityKey key)
 
             if (key.IsSupportedAlgorithm(SecurityAlgorithms.Aes256KW))
             {
+                if (key.KeySize != 256)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(key), SR.GetResourceString(SR.ID0283));
```
Comment (kevinchalet) [inline src/OpenIddict.Server/OpenIddictServerBuilder.cs]: Let's use `InvalidOperationException` to match the other exceptions thrown from this method.

File: test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs [hunk 4]
```diff
@@ -181,6 +181,32 @@ public void AddEncryptionKey_EncryptingKeyIsCorrectlyAdded()
...
+        public void AddEncryptionKey_ThrowsExceptionWhenSymmetricKeyIsTooShort()
+        {
+            // Arrange
+            var services = CreateServices();
+            var builder = CreateBuilder(services);
+
+            // Act and assert
+            var key = Mock.Of<SecurityKey>(mock => mock.KeySize == 128 && mock.IsSupportedAlgorithm(SecurityAlgorithms.Aes256KW));
+            var exception = Assert.Throws<ArgumentOutOfRangeException>(() => builder.AddEncryptionKey(key));
+            Assert.Equal("key", exception.ParamName);
```
Comment (kevinchalet) [inline test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs]: We'll also need to check the actual error message.

File: test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs [hunk 5]
```diff
@@ -181,6 +181,32 @@ public void AddEncryptionKey_EncryptingKeyIsCorrectlyAdded()
...
+        public void AddEncryptionKey_ThrowsExceptionWhenSymmetricKeyIsTooLong()
+        {
+            // Arrange
+            var services = CreateServices();
+            var builder = CreateBuilder(services);
+
+            // Act and assert
+            var key = Mock.Of<SecurityKey>(mock => mock.KeySize == 384 && mock.IsSupportedAlgorithm(SecurityAlgorithms.Aes256KW));
+            var exception = Assert.Throws<ArgumentOutOfRangeException>(() => builder.AddEncryptionKey(key));
+            Assert.Equal("key", exception.ParamName);
```
Comment (kevinchalet) [inline test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs]: Here too.

**Review by kevinchalet — CHANGES_REQUESTED** [review]

**Comment by kevinchalet** [comment]
Thanks for your PR!

**Comment by NoahStahl** [comment]
@kevinchalet Thanks for the feedback. I've added the suggested changes.

File: src/OpenIddict.Server/OpenIddictServerBuilder.cs [hunk 6]
```diff
@@ -157,7 +157,9 @@ public OpenIddictServerBuilder AddEncryptionCredentials(EncryptingCredentials cr
         /// <summary>
         /// Registers an encryption key.
         /// </summary>
-        /// <param name="key">The security key.</param>
+        /// <param name="key">The security key. If a SymmetricSecurityKey, create using a 32-character
```
Comment (kevinchalet) [inline src/OpenIddict.Server/OpenIddictServerBuilder.cs]: Technically, this comment is not 100% exact, as the check doesn't only apply to `SymmetricSecurityKey` instances (e.g you could have a `JsonWebKey` instance representing a symmetric key). Let's revert this change, please.

File: test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs [hunk 7]
```diff
@@ -181,6 +181,32 @@ public void AddEncryptionKey_EncryptingKeyIsCorrectlyAdded()
...
+        public void AddEncryptionKey_ThrowsExceptionWhenSymmetricKeyIsTooShort()
+        {
+            // Arrange
+            var services = CreateServices();
+            var builder = CreateBuilder(services);
+
+            // Act and assert
+            var key = Mock.Of<SecurityKey>(mock => mock.KeySize == 128 && mock.IsSupportedAlgorithm(SecurityAlgorithms.Aes256KW));
+            var exception = Assert.Throws<InvalidOperationException>(() => builder.AddEncryptionKey(key));
+            Assert.Equal("Provided symmetric key was incorrect size. Expected 256 bits, received 128.", exception.Message);
```
Comment (kevinchalet) [inline test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs]: Please use `SR.FormatID0283(256, 128)` here.

File: test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs [hunk 8]
```diff
@@ -181,6 +181,32 @@ public void AddEncryptionKey_EncryptingKeyIsCorrectlyAdded()
...
+        public void AddEncryptionKey_ThrowsExceptionWhenSymmetricKeyIsTooLong()
+        {
+            // Arrange
+            var services = CreateServices();
+            var builder = CreateBuilder(services);
+
+            // Act and assert
+            var key = Mock.Of<SecurityKey>(mock => mock.KeySize == 384 && mock.IsSupportedAlgorithm(SecurityAlgorithms.Aes256KW));
+            var exception = Assert.Throws<InvalidOperationException>(() => builder.AddEncryptionKey(key));
+            Assert.Equal("Provided symmetric key was incorrect size. Expected 256 bits, received 384.", exception.Message);
```
Comment (kevinchalet) [inline test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs]: Same here.

**Review by kevinchalet — APPROVED** [review]

**Comment by NoahStahl** [comment]
Thanks for additional comments Kevin, changes made. Sorry for additional commit, got confused there for a moment.

**Comment by kevinchalet** [comment]
Merged, thanks for your contribution!

**Pull request merged**